File: hello-cpp/client/src/core/upload_manager.cpp
#include "core/upload_manager.h"
#include "../../include/common_types.h"
#include "util/performance_monitor.h"
#include <chrono>
#include <nlohmann/json.hpp>
#include <spdlog/spdlog.h>
#include <thread>

namespace audio_stream {

UploadManager::UploadManager(std::shared_ptr<WebSocketClient> client,
                             std::shared_ptr<ErrorHandler> errorHandler)
    : client_(client), errorHandler_(errorHandler), responseReceived_(false),
      responseTimeoutMs_(5000) {
  // Message handling is now done by main.cpp message router
}

std::string UploadManager::uploadFile(const std::string &filePath) {
  spdlog::info("Starting upload of file: {}", filePath);

  // Validate file exists and is readable
  if (!fileManager_.openForReading(filePath)) {
    if (errorHandler_) {
      errorHandler_->handleFileIOError("Cannot open file for reading",
                                       filePath);
    }
    return "";
  }
  fileManager_.closeReader(); // Close immediately, will reopen later

  // Generate unique stream ID
  currentStreamId_ = streamIdGenerator_.generateStreamId();
  spdlog::info("Generated stream ID: {}", currentStreamId_);

  // Start performance monitoring
  performanceMonitor_.startUpload();

  try {
    // Step 1: Send START message
    if (!sendStartMessage(currentStreamId_)) {
      if (errorHandler_) {
        errorHandler_->reportError(ErrorHandler::ErrorType::PROTOCOL_ERROR,
                                   "Failed to send START message",
                                   "Stream ID: " + currentStreamId_, false);
      }
      return "";
    }

    // Step 2: Send file chunks
    if (!sendFileChunks(filePath)) {
      if (errorHandler_) {
        errorHandler_->reportError(ErrorHandler::ErrorType::PROTOCOL_ERROR,
                                   "Failed to send file chunks",
                                   "File: " + filePath, false);
      }
      return "";
    }

    // Step 3: Send STOP message
    if (!sendStopMessage(currentStreamId_)) {
      if (errorHandler_) {
        errorHandler_->reportError(ErrorHandler::ErrorType::PROTOCOL_ERROR,
                                   "Failed to send STOP message",
                                   "Stream ID: " + currentStreamId_, false);
      }
      return "";
    }

    // End performance monitoring
    performanceMonitor_.endUpload(fileManager_.getFileSize());

    spdlog::info("Successfully uploaded file: {} with stream ID: {}", filePath,
                 currentStreamId_);
    return currentStreamId_;

  } catch (const std::exception &e) {
    if (errorHandler_) {
      errorHandler_->reportError(ErrorHandler::ErrorType::PROTOCOL_ERROR,
                                 "Exception during upload: " +
                                     std::string(e.what()),
                                 "File: " + filePath, false);
    }
    return "";
  }
}

bool UploadManager::sendStartMessage(const std::string &streamId) {
  spdlog::debug("Sending START message for stream: {}", streamId);

  StartMessage startMsg;
  startMsg.streamId = streamId;

  nlohmann::json j;
  j["type"] = startMsg.type;
  j["streamId"] = startMsg.streamId;
  std::string jsonMessage = j.dump();

  responseReceived_ = false;
  client_->sendTextMessage(jsonMessage);

  // Wait for STARTED response with timeout
  waitForResponse("STARTED", responseTimeoutMs_);

  if (!responseReceived_) {
    if (errorHandler_) {
      errorHandler_->handleTimeoutError(
          "No response received for START message", responseTimeoutMs_);
    }
    return false;
  }

  try {
    nlohmann::json responseJson = nlohmann::json::parse(lastResponse_);
    if (responseJson["type"] == "STARTED") {
      spdlog::info("Received STARTED response: {}",
                   responseJson["message"].get<std::string>());
      return true;
    } else if (responseJson["type"] == "ERROR") {
      std::string errorMsg = responseJson.contains("message")
                                 ? responseJson["message"].get<std::string>()
                                 : "Unknown error";
      return handleProtocolError("Server error in START: " + errorMsg,
                                 "START message");
    } else {
      return handleProtocolError("Unexpected response type: " +
                                     responseJson["type"].get<std::string>(),
                                 "Expected 'started'");
    }
  } catch (const std::exception &e) {
    return handleProtocolError("Failed to parse STARTED response: " +
                                   std::string(e.what()),
                               "JSON parsing");
  }
}

bool UploadManager::sendFileChunks(const std::string &filePath) {
  spdlog::debug("Sending file chunks for: {}", filePath);

  // Open file for reading
  if (!fileManager_.openForReading(filePath)) {
    if (errorHandler_) {
      errorHandler_->handleFileIOError("Failed to open file for reading",
                                       filePath);
    }
    return false;
  }

  size_t totalSize = fileManager_.getFileSize();
  size_t bytesUploaded = 0;

  spdlog::info("File size: {} bytes, estimated chunks: {}", totalSize,
               chunkManager_.calculateChunkCount(totalSize));

  try {
    // Read and send chunks
    while (fileManager_.hasMoreData()) {
      std::vector<uint8_t> chunk;
      size_t bytesRead = fileManager_.readChunk(chunk);

      if (bytesRead == 0) {
        break; // End of file
      }

      // Send chunk as binary message
      client_->sendBinaryMessage(chunk);

      bytesUploaded += bytesRead;

      // Call progress callback if set
      if (progressCallback_) {
        progressCallback_(bytesUploaded, totalSize);
      }

      spdlog::debug("Sent chunk: {} bytes (total: {}/{})", bytesRead,
                    bytesUploaded, totalSize);
    }

    fileManager_.closeReader();

    spdlog::info("Finished sending {} bytes in chunks", bytesUploaded);
    return true;

  } catch (const std::exception &e) {
    fileManager_.closeReader();
    if (errorHandler_) {
      errorHandler_->reportError(ErrorHandler::ErrorType::FILE_IO_ERROR,
                                 "Exception while reading file chunks: " +
                                     std::string(e.what()),
                                 "File: " + filePath, false);
    }
    return false;
  }
}

bool UploadManager::handleProtocolError(const std::string &message,
                                        const std::string &context) {
  if (errorHandler_) {
    errorHandler_->handleProtocolError(message, context);
  } else {
    spdlog::error("Protocol error: {} (Context: {})", message, context);
  }
  return false;
}

bool UploadManager::sendStopMessage(const std::string &streamId) {
  spdlog::debug("Sending STOP message for stream: {}", streamId);

  StopMessage stopMsg;
  stopMsg.streamId = streamId;

  nlohmann::json j;
  j["type"] = stopMsg.type;
  j["streamId"] = stopMsg.streamId;
  std::string jsonMessage = j.dump();

  responseReceived_ = false;
  client_->sendTextMessage(jsonMessage);

  // Wait for STOPPED response with timeout
  waitForResponse("STOPPED", responseTimeoutMs_);

  if (!responseReceived_) {
    if (errorHandler_) {
      errorHandler_->handleTimeoutError("No response received for STOP message",
                                        responseTimeoutMs_);
    }
    return false;
  }

  try {
    nlohmann::json responseJson = nlohmann::json::parse(lastResponse_);
    if (responseJson["type"] == "STOPPED") {
      spdlog::info("Received STOPPED response: {}",
                   responseJson["message"].get<std::string>());
      return true;
    } else if (responseJson["type"] == "ERROR") {
      std::string errorMsg = responseJson.contains("message")
                                 ? responseJson["message"].get<std::string>()
                                 : "Unknown error";
      return handleProtocolError("Server error in STOP: " + errorMsg,
                                 "STOP message");
    } else {
      return handleProtocolError("Unexpected response type: " +
                                     responseJson["type"].get<std::string>(),
                                 "Expected 'STOPPED'");
    }
  } catch (const std::exception &e) {
    return handleProtocolError("Failed to parse STOPPED response: " +
                                   std::string(e.what()),
                               "JSON parsing");
  }
}

void UploadManager::waitForResponse(const std::string &expectedType,
                                    int timeoutMs) {
  auto startTime = std::chrono::steady_clock::now();
  auto timeout = std::chrono::milliseconds(timeoutMs);

  while (!responseReceived_) {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));

    auto elapsed = std::chrono::steady_clock::now() - startTime;
    if (elapsed > timeout) {
      spdlog::warn("Timeout waiting for {} response", expectedType);
      break;
    }
  }
}

void UploadManager::setProgressCallback(
    std::function<void(size_t, size_t)> callback) {
  progressCallback_ = callback;
}

PerformanceMetrics UploadManager::getPerformanceMetrics() const {
  return performanceMonitor_.getMetrics();
}

void UploadManager::handleServerResponse(const std::string &message) {
  spdlog::debug("Received server response: {}", message);
  lastResponse_ = message;
  responseReceived_ = true;
}

} // namespace audio_stream

File: hello-java/audio-stream-client/src/main/java/org/feuyeux/mmap/audio/client/core/UploadManager.java
package org.feuyeux.mmap.audio.client.core;

import org.feuyeux.mmap.audio.client.util.ErrorHandler;
import org.feuyeux.mmap.audio.client.util.PerformanceMonitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;

/**
 * Upload manager for orchestrating file upload workflow.
 * Handles the complete upload process: START -> chunks -> STOP.
 * Matches the C++ UploadManager interface.
 */
public class UploadManager {
    private static final Logger logger = LoggerFactory.getLogger(UploadManager.class);
    private static final int MESSAGE_PAUSE_MS = 500;
    private static final int DEFAULT_UPLOAD_DELAY_MS = 10;
    private static final int CHUNK_SIZE = 65536; // 64KB

    private final WebSocketClient client;
    private final ErrorHandler errorHandler;
    private final PerformanceMonitor performanceMonitor;

    private BiConsumer<Long, Long> progressCallback;

    private int uploadDelayMs;

    /**
     * Create an upload manager.
     *
     * @param client       WebSocket client
     * @param errorHandler error handler (optional)
     */
    public UploadManager(WebSocketClient client, ErrorHandler errorHandler) {
        this(client, errorHandler, new PerformanceMonitor());
    }

    /**
     * Create an upload manager with all dependencies.
     *
     * @param client             WebSocket client
     * @param errorHandler       error handler
     * @param performanceMonitor performance monitor
     */
    public UploadManager(WebSocketClient client, ErrorHandler errorHandler,
                         PerformanceMonitor performanceMonitor) {
        this.client = client;
        this.errorHandler = errorHandler;
        this.performanceMonitor = performanceMonitor;
        this.uploadDelayMs = DEFAULT_UPLOAD_DELAY_MS;
    }


    /**
     * Upload a file to the server.
     *
     * @param filePath path to the file to upload
     * @return generated stream ID if successful, empty string if failed
     */
    public String uploadFile(Path filePath) {
        if (!Files.exists(filePath)) {
            if (errorHandler != null) {
                errorHandler.reportError(ErrorHandler.ErrorType.FILE_IO_ERROR,
                        "File not found", filePath.toString(), false);
            }
            logger.error("File not found: {}", filePath);
            return "";
        }

        try {
            long fileSize = Files.size(filePath);
            String streamId = generateStreamId();

            logger.info("Starting upload - File: {}, Size: {} bytes, StreamId: {}",
                    filePath.getFileName(), fileSize, streamId);

            performanceMonitor.startUpload();

            // Consume CONNECTED message if present
            try {
                String connectedMsg = client.waitForTextMessage(1000);
                if (connectedMsg != null && (connectedMsg.contains("\"type\":\"CONNECTED\"") || connectedMsg.contains("\"type\":\"connected\""))) {
                    logger.debug("Consumed CONNECTED message");
                }
            } catch (Exception e) {
                // Ignore if no CONNECTED message
            }

            // Send START message
            if (!sendStartMessage(streamId)) {
                return "";
            }

            TimeUnit.MILLISECONDS.sleep(MESSAGE_PAUSE_MS);

            // Send file chunks
            if (!sendFileChunks(filePath, fileSize)) {
                return "";
            }

            TimeUnit.MILLISECONDS.sleep(MESSAGE_PAUSE_MS);

            // Send STOP message
            if (!sendStopMessage(streamId)) {
                return "";
            }

            performanceMonitor.endUpload(fileSize);

            logger.info("Upload completed successfully with stream ID: {}", streamId);
            return streamId;

        } catch (IOException e) {
            if (errorHandler != null) {
                errorHandler.reportError(ErrorHandler.ErrorType.FILE_IO_ERROR,
                        "Failed to read file: " + e.getMessage(), filePath.toString(), false);
            }
            logger.error("Upload failed for file: {}", filePath, e);
            return "";
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            if (errorHandler != null) {
                errorHandler.reportError(ErrorHandler.ErrorType.PROTOCOL_ERROR,
                        "Upload interrupted", filePath.toString(), false);
            }
            logger.error("Upload interrupted for file: {}", filePath, e);
            return "";
        }
    }

    // Private helper methods

    /**
     * Send START message to begin streaming.
     *
     * @param streamId stream identifier
     * @return true if successful
     */
    private boolean sendStartMessage(String streamId) {
        try {
            String message = String.format("{\"type\":\"START\",\"streamId\":\"%s\"}", streamId);
            client.sendTextMessage(message);
            logger.info("Sent START message for stream: {}", streamId);
            return true;
        } catch (Exception e) {
            if (errorHandler != null) {
                errorHandler.reportError(ErrorHandler.ErrorType.PROTOCOL_ERROR,
                        "Failed to send START message: " + e.getMessage(), streamId, false);
            }
            logger.error("Failed to send START message", e);
            return false;
        }
    }

    /**
     * Send file chunks to the server.
     *
     * @param filePath path to the file
     * @param fileSize size of the file
     * @return true if successful
     */
    private boolean sendFileChunks(Path filePath, long fileSize) throws IOException, InterruptedException {
        byte[] fileData = Files.readAllBytes(filePath);
        int offset = 0;
        int totalChunks = 0;
        long totalBytesTransferred = 0;

        while (offset < fileData.length) {
            int chunkLength = Math.min(CHUNK_SIZE, fileData.length - offset);
            byte[] chunk = new byte[chunkLength];
            System.arraycopy(fileData, offset, chunk, 0, chunkLength);

            client.sendBinaryMessage(chunk);
            totalBytesTransferred += chunkLength;
            totalChunks++;

            // Call progress callback if set
            if (progressCallback != null) {
                progressCallback.accept(totalBytesTransferred, fileSize);
            }

            if (uploadDelayMs > 0) {
                TimeUnit.MILLISECONDS.sleep(uploadDelayMs);
            }

            offset += chunkLength;

            if (totalChunks % 100 == 0) {
                logger.info("Upload progress: {} / {} bytes ({}%)",
                        totalBytesTransferred, fileSize,
                        String.format("%.1f", totalBytesTransferred * 100.0 / fileSize));
            }
        }

        logger.info("Sent {} chunks ({} bytes)", totalChunks, totalBytesTransferred);
        return true;
    }

    /**
     * Send STOP message to end streaming.
     *
     * @param streamId stream identifier
     * @return true if successful
     */
    private boolean sendStopMessage(String streamId) {
        try {
            String message = "{\"type\":\"STOP\"}";
            client.sendTextMessage(message);
            logger.info("Sent STOP message");
            return true;
        } catch (Exception e) {
            if (errorHandler != null) {
                errorHandler.reportError(ErrorHandler.ErrorType.PROTOCOL_ERROR,
                        "Failed to send STOP message: " + e.getMessage(), streamId, false);
            }
            logger.error("Failed to send STOP message", e);
            return false;
        }
    }


    /**
     * Generate a short stream ID.
     *
     * @return stream ID in format "stream-{short-uuid}"
     */
    private static String generateStreamId() {
        String uuid = UUID.randomUUID().toString().substring(0, 8);
        String streamId = "stream-" + uuid;
        logger.debug("Generated stream ID: {}", streamId);
        return streamId;
    }
}
